aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -75,44 +77,39 @@ void index_select_add<float>(const Tensor &select_indices,
                              const Tensor &src,
                              Tensor &output,
                              const Tensor& offsets) {
+
   int64_t ddim = src.size(1);
+  auto src_stride0 = src.stride(0);
+  auto src_stride1 = src.stride(1);
+  auto output_stride0 = output.stride(0);
+  auto output_stride1 = output.stride(1);
+  auto select_numel = select_indices.numel();
+
   auto src_data = src.data<float>();
   auto select_indices_data = select_indices.data<int64_t>();
   auto output_data = output.data<float>();
 
-  if (isFastPathIndexSelect(src, output)) {
-    auto accessor = offsets.accessor<int64_t, 1>();
-    std::vector<int> lengths;
-
-    int64_t lower = accessor[0];
-    for (int64_t i = 1; i < offsets.numel(); ++i) {
-      lengths.push_back(accessor[i] - lower);
-      lower = accessor[i];
-    }
-    lengths.push_back(select_indices.numel() - lower);
-
-    caffe2::EmbeddingLookup(
-      /*block_size=*/ddim,
-      /*output_size=*/lengths.size(),
-      /*index_size=*/select_indices.numel(),
-      /*data_size=*/src.size(0),
-      /*input=*/src_data,
-      /*indices=*/select_indices_data,
-      /*lengths=*/lengths.data(),
-      /*weights=*/nullptr,
-      /*scale_bias=*/nullptr,
-      /*normalize_by_lengths=*/false,
-      /*out=*/output_data
-    );
+  if ((src_stride1 == 1) && (output_stride1 == 1)) {
+    auto offset_data = offsets.data<int64_t>();
+    auto offset_numel = offsets.numel();
+    at::parallel_for(0, offset_numel, 0, [&](int64_t start, int64_t end) {
+      for(auto mb = start; mb < end; mb++) {
+        auto select_off_start = offset_data[mb];
+        auto select_off_end = (mb < (offset_numel - 1) ? offset_data[mb + 1] : select_numel);
+        auto output_block = output_data + output_stride0 * mb;
+        for(auto s = select_off_start; s < select_off_end; s++) {
+          auto src_offset = src_stride0 * select_indices_data[s];
+          PRAGMA_SIMD
```

Comment:
`PRAGMA_SIMD` is more Intel specific (Cilk? Correct me if I am wrong). I am wondering what is the experimental setup to get the ~107.54 ms inference time for the DLRM benchmark (Which Intel architecture, which compiler)? Can you show the similar performance with GCC? Thanks! cc @dmudiger, @jspark1105.

Reply:
PRAGMA_SIMD is provided by the Pytorch parallel framework, it's defined as below and can be compiled by GCC

#ifdef _MSC_VER
#define PRAGMA(P)         __pragma(P)
# if _MSC_VER < 1920
// MSVC < 2019 doesn't support loop pragmas.
#  define PRAGMA_IVDEP    // Noop
#  define PRAGMA_SIMD     // Noop
# else
#  define PRAGMA_IVDEP    PRAGMA(loop(ivdep))
#  define PRAGMA_SIMD     PRAGMA(omp simd)
# endif
#else
#define PRAGMA(P)         _Pragma(#P)
#define PRAGMA_IVDEP      PRAGMA(ivdep)
#define PRAGMA_SIMD       PRAGMA(omp simd)
#endif

To get 107ms performance, need to combine the codes modified at #23055 and #23057. For just this function, it can improved ~40ms

Reply:
Thanks for pointing it out! This makes sense as `PRAGMA_SIMD` can be `pragma simd` or `pragma omp simd` for different platforms. We recently made an update to the `caffe2::EmbeddingLookup` to directly support the offsets instead of lengths (https://github.com/pytorch/pytorch/pull/24945), so it should remove some overhead. As the next step, we plan to add the AVX512 support for this kernel and use JIT to generate the kernel to achieve more performance speedup soon.

Regarding the parallelism support inside EmbeddingBag (making index_select_add fast path parallel), we might need some further discussion. Normally in many applications using this operators, we tend to have the parallelism in the inter-op level instead of intra-op level. I wonder the performance you achieved with single thread using DLRM benchmark inference on Intel SKX-8180.

Reply:
yeah, I will have a look into #24945 and have a compare on this. It's very nice to have a whole AVX512 support for this kernel and when complete please let me know and I can have also try on this. 
For parallelism support, may we have some options or configuration to tune the parallel level for the inter-op or intra-op ? 
Thanks